File: gen_map.pl
#!/usr/bin/perl

use Time::Piece;

# Init Zone hash
# $zone{zone_loc}{NAME} = "City Name" ;
# $zone{zone_loc}{COLOR} = color code/#xxxxxx ;
# $zone{zone_loc}{CURRENT} = 1 ;
my %zone = ();

# caves of qud save location & specific game save location
my $savedir = "C:\\Users\\owner\\AppData\\LocalLow\\Freehold\ Games\\CavesOfQud";
my $saveUID = "0469531a-048d-4549-b26d-88034706eeac";

# output file
my $html_file = 'parsang_map.html';

# csv list of cities (in Save Dir)
my $locations_csv = 'cities.csv';


###############################################################################
# main loop
###############################################################################

# &read_cache_dir();

while(1) {

    my $current_modified = (stat $savedir . "\\Player.log")[9];

    if( $current_modified-$prev_modified > 15 ) {

        print("Detected file change...Regenerating map $current_modified\n");

        &read_player_log();

#        &add_locations();

        &gen_html_output();

        $prev_modified = $current_modified;

    };

    # limit disk thrashing
    sleep(10);

}

exit ;


###############################################################################
# Subs
###############################################################################

###############################################################################
# Read filenames in zone cache dir
# Add locatioin to zone hash
# Set any visited locations to lightgrey
###############################################################################
sub read_cache_dir() {

    my $zone_cache_dir = $savedir . "\\Saves\\$saveUID\\ZoneCache";

    # read zone cache dir
    opendir(CACHEDIR,$zone_cache_dir) || die "Can't opendir $zone_cache_dir: $!";
        my @zonecache = readdir(CACHEDIR);
    close CACHEDIR;

    foreach(@zonecache) {
        $_ =~ s/JoppaWorld\.//g;
        $_ =~ s/\.zone\.gz//g;
        # print("$_\n");
        $zone{$_}{COLOR} = 'lightgrey';
    }

}

###############################################################################
# Parse Player.log file
# Look for zones that are Thawing/Building
# Set any found locations to grey
# Set last found location to CURRENT
###############################################################################
sub read_player_log(){
    # Read current Player.log
    # tail -n 1000 -f Player.log | awk -F. '/INFO - Thawing|INFO - Building/{print $2 "." $4 " " $3 "." $5 " :" 10-$6}' 
    my $player_log_file = $savedir . "\\Player.log";

    # unset old CURRENT
    foreach $loc (keys %zone) {
        if (exists $zone{$loc}{CURRENT}){
            delete $zone{$loc}{CURRENT} ;
        }
    }

    open(PLAYERLOG, "<", $player_log_file) || die "Can't open file $player_log_file: $!";
        while(<PLAYERLOG>){
            if($_ =~ m/INFO - Finished \'Thawing|INFO - Finished \'Building/){
                $_ =~ /(\d+.\d+.\d+.\d+.\d+)/;
                $current_location=$1;
                $zone{$1}{COLOR} = 'grey';
            }
        }

        # Set the last read location as Current: color magenta
        $zone{$current_location}{COLOR} = 'magenta';
        $zone{$current_location}{CURRENT} = 1 ;

	printf("Current Location: %s\n",$current_location);
    close(PLAYERLOG);

}

###############################################################################
# Add city/sites of discovered locations
# Read file with found cities
#
# City list format (Location, Color, Name)
#  11.22.1.1.10,#554f97,Joppa
#
#  ?TODO? Need to parse Prmary.sav and pull locations from there
###############################################################################
sub add_locations() {

    my $filename = $savedir . "\\" . $locations_csv ;

    open(FH, "<", $filename) || die "Can't open file $filename: $!";

    while(<FH>){
        # ($loc, $color, $name) = split(/,/,$_);

        next unless ($_ =~ /^(\d{1,2}\.\d{1,2}\.\d\.\d\.\d{1,2}),(.+),(.+)$/ ) ;
        $zone{&trim($1)}{NAME} = &trim($3) ;
        $zone{&trim($1)}{COLOR} = &trim($2);
    }

    close(FH) ;

}


#
# Write HTML File
#
sub gen_html_output(){
    open(HTMLOUT, '>', $html_file) or die $!;

    &gen_html_header(); 
    &gen_html_table();
    &gen_html_footer();

    close(HTMLOUT);
}

#
# Trim leading/trailing spaces from string
#
sub  trim { my $s = shift; $s =~ s/^\s+|\s+$//g; return $s };

sub gen_html_header() {

    print HTMLOUT ("<!DOCTYPE html>\n");
    print HTMLOUT ("<html>\n");
    print HTMLOUT ("<head>\n");

    print HTMLOUT ("\t<title>CavesOfQud Parsang Map</title>\n");
    print HTMLOUT ("\t<meta http-equiv=\"refresh\" content=\"30\">\n");

    print HTMLOUT ("\t<style>\n");
    print HTMLOUT ("\t\ttable {\n");
    print HTMLOUT ("\t\t\tfont-family: arial, sans-serif;\n");
    print HTMLOUT ("\t\t\tborder-collapse: collapse;\n");
    print HTMLOUT ("\t\t\twidth: 100%;\n");
    print HTMLOUT ("\t\t}\n");

    print HTMLOUT ("\t\tth {\n");
    print HTMLOUT ("\t\t\tbackground-color: lightslategrey;\n");
    print HTMLOUT ("\t\t\tborder: 4px solid #d70513;\n");
    print HTMLOUT ("\t\t\tposition: sticky;\n");
    print HTMLOUT ("\t\t\ttop: 0;\n");
    print HTMLOUT ("\t\t\tleft: 0;\n");
    print HTMLOUT ("\t\t\ttext-align: center;\n");
    print HTMLOUT ("\t\t\tpadding: 8px;\n");
    print HTMLOUT ("\t\t}\n");

    print HTMLOUT ("\t\ttd {\n");
    print HTMLOUT ("\t\t\tborder: 1px solid #dddddd;\n");
    print HTMLOUT ("\t\t\ttext-align: center;\n");
    print HTMLOUT ("\t\t\ttext-overflow: ellipsis;\n");
    print HTMLOUT ("\t\t\tpadding: 8px;\n");
    print HTMLOUT ("\t\t}\n");

    print HTMLOUT ("\t\ttr.border-bottom {\n");
    print HTMLOUT ("\t\t\tborder-bottom: solid 4px #d70513;\n");
    print HTMLOUT ("\t\t}\n");

    print HTMLOUT ("\t\ttr.border-top {\n");
    print HTMLOUT ("\t\t\tborder-top: solid 4px #d70513;\n");
    print HTMLOUT ("\t\t}\n");

    print HTMLOUT ("\t\ttd.border-left {\n");
    print HTMLOUT ("\t\t\tborder-left: solid 4px #d70513;\n");
    print HTMLOUT ("\t\t}\n");

    print HTMLOUT ("\t\ttd.border-right {\n");
    print HTMLOUT ("\t\t\tborder-right: solid 4px #d70513;\n");
    print HTMLOUT ("\t\t}\n");

    print HTMLOUT ("\t\ttd.border-top {\n");
    print HTMLOUT ("\t\t\tborder-top: solid 4px #d70513;\n");
    print HTMLOUT ("\t\t}\n");

    print HTMLOUT ("\t\ttd.border-bottom {\n");
    print HTMLOUT ("\t\t\tborder-bottom: solid 4px #d70513;\n");
    print HTMLOUT ("\t\t}\n");

    print HTMLOUT ("\t\ttd.current-location {\n");
    print HTMLOUT ("\t\t\tborder: solid 8px #f817d6;\n");
    print HTMLOUT ("\t\t}\n");

    print HTMLOUT ("\t</style>\n");
    print HTMLOUT ("</head>\n");

    print HTMLOUT ("\n");

}


sub gen_html_table($) {

    my $tr_class;
    my $td_class;
    my $td_bgcolor;
    my $td_str;
    my $zone_depth = 10 ; 

    print HTMLOUT ("<body>\n");
    print HTMLOUT ("\t<center><h1>Caves Of Qud Parsang Map</h1></center>\n");

    printf HTMLOUT ("Generated: %s\n", localtime->datetime);

    print HTMLOUT ("\t<table>\n");

    # output map header
    print HTMLOUT ("\t\t<tr>\n");
    print HTMLOUT ("\t\t\t<th>&nbsp;</th>");
    for ( my $parsang_x = 0; $parsang_x < 80; $parsang_x++) {
        for ( my$zone_x = 0; $zone_x <= 2; $zone_x++) {
            printf HTMLOUT ("<th>%s.%s</th>",$parsang_x,$zone_x);
        }
    }
    print HTMLOUT ("\n");

    # output map data
    for ( my $parsang_y = 0; $parsang_y < 25; $parsang_y++) {
        for ( my $zone_y = 0; $zone_y <= 2; $zone_y++) {

            # draw border for the 3x3 grid
            if ($zone_y == 0) {
                $tr_class = "class=\"border-top\"" ;
            } else {
                $tr_class = "" ;
            }

            # open data row
            printf HTMLOUT ("\t\t<tr %s>\n", $tr_class);

            # output row index
            printf HTMLOUT ("\t\t\t<th>%s.%s</th>",$parsang_y, $zone_y);

            # output row data
            for ( $parsang_x = 0; $parsang_x < 80; $parsang_x++ ) {
                for ( $zone_x = 0; $zone_x <= 2; $zone_x++) {

                    # set border based on column
                    if ( $zone_x == 0 ) {
                        $td_class = "class=\"border-left\"" ; 
                    } else {
                        $td_class = "" ; 
                    }

                    # set bgcolor if visited
                    # $zone{'1.1.1.1.10'}{'visited'} = 'color spec';
                    my $zone_loc = join('.',$parsang_x,$parsang_y,$zone_x,$zone_y,$zone_depth);
                    # print("$zone_loc\n");

                    # populate City locations
                    if ( exists($zone{$zone_loc}{NAME})) {
                        $td_str = $zone{$zone_loc}{NAME} ;
                    } else {
                        $td_str = '&nbsp;' ;
                    }

                    # show current location
                    if ( exists($zone{$zone_loc}{COLOR})) {
                        if (exists($zone{$zone_loc}{CURRENT})) {
                            $td_class = "class=\"current-location\"" ;
                            # $td_str = "$parsang_x.$zone_x $parsang_y.$zone_y" if $td_str == '&nbsp;';
                        }
                        $td_bgcolor = "bgcolor=" . $zone{$zone_loc}{COLOR} ;
                    } else {
                        $td_bgcolor = '' ;
                    }

                    printf HTMLOUT ("<td %s %s>%s</td>", $td_class, $td_bgcolor, $td_str);
                }
            }
            printf HTMLOUT ("\n");

            # end data row
            print HTMLOUT ("\t\t</tr>\n");
            printf HTMLOUT ("\n");
        }
    }

    printf HTMLOUT ("\n");

    print HTMLOUT ("\t</table>\n");
    print HTMLOUT ("</body>\n");

    print HTMLOUT ("\n");

}


sub gen_html_footer() {

    print HTMLOUT ("</html>\n");
    print HTMLOUT ("\n");

}

File: gen_map.py

import os
import time
import re
from datetime import datetime

# --- Configuration ---
# Define the base directory for Caves of Qud game saves and logs.
# This path is specific to a Windows user.
# IMPORTANT: Adjust this path to your actual Caves of Qud save directory.
SAVE_DIR = "C:\\Users\\owner\\AppData\\LocalLow\\Freehold Games\\CavesOfQud"

# Define the unique ID for a specific game save.
# You can find this in your SAVE_DIR/Saves/ directory (it's the folder name).
SAVE_UID = "0469531a-048d-4549-b26d-88034706eeac"

# Define the name of the output HTML file for the map.
HTML_FILE = 'parsang_map.html'

# Define the name of the CSV file containing known city locations.
# This file should be located in the SAVE_DIR.
LOCATIONS_CSV = 'cities.csv'

# --- Global Data Structure (using a dictionary for zones) ---
# A dictionary to store information about game zones.
# Keys will be zone coordinates (e.g., "1.1.1.1.10").
# Values will be dictionaries containing 'name', 'color', and 'current' status.
zones = {}

# --- Utility Functions ---

def trim(s: str) -> str:
    """
    Trims leading and trailing whitespace from a string.
    """
    return s.strip()

# --- Core Logic Functions ---

# This function is equivalent to Perl's 'read_cache_dir' and is currently unused
# in the main loop, mirroring the commented-out status in the original Perl.
# It's included here for completeness if you decide to use it later.
def read_zone_cache_dir():
    """
    Reads filenames from the game's ZoneCache directory.
    Marks these zones as 'lightgrey' (visited) in the global 'zones' dictionary.
    """
    zone_cache_dir = os.path.join(SAVE_DIR, "Saves", SAVE_UID, "ZoneCache")

    if not os.path.exists(zone_cache_dir):
        print(f"Warning: ZoneCache directory not found: {zone_cache_dir}")
        return

    try:
        # List all files and directories in the ZoneCache.
        # Filter for actual zone files.
        zone_files = [f for f in os.listdir(zone_cache_dir) if f.endswith(".zone.gz")]

        for filename in zone_files:
            # Remove "JoppaWorld." prefix and ".zone.gz" suffix to get the pure zone coordinate.
            zone_loc = filename.replace("JoppaWorld.", "").replace(".zone.gz", "")
            zones[zone_loc] = zones.get(zone_loc, {}) # Ensure the inner dict exists
            zones[zone_loc]['color'] = 'lightgrey'
    except OSError as e:
        print(f"Error reading ZoneCache directory {zone_cache_dir}: {e}")

def read_player_log():
    """
    Parses the Player.log file to find the player's current and recently visited zones.
    Updates the global 'zones' dictionary with color and current status.
    """
    player_log_file = os.path.join(SAVE_DIR, "Player.log")

    if not os.path.exists(player_log_file):
        print(f"Error: Player.log not found at {player_log_file}")
        return

    # Unset the 'current' flag for all previously known zones.
    for loc_data in zones.values():
        if 'current' in loc_data:
            del loc_data['current']

    current_location = None
    # Regex to capture the zone coordinate from log lines indicating zone loading.
    # e.g., "INFO - Finished 'Thawing Zone 1.1.1.1.10'"
    log_pattern = re.compile(r"INFO - Finished '(?:Thawing|Building) \b.+\.(\d+\.\d+\.\d+\.\d+\.\d+)'")

    try:
        with open(player_log_file, 'r', encoding='utf-8', errors='ignore') as f:
            # Read all lines and process them to find the last known location.
            # This ensures we get the *most recent* location even if the log is large.
            for line in f:
                match = log_pattern.search(line)
                if match:
                    zone_loc = match.group(1)
                    zones[zone_loc] = zones.get(zone_loc, {}) # Ensure the inner dict exists
                    zones[zone_loc]['color'] = 'grey' # Mark as visited
                    current_location = zone_loc # Update the last found location

        if current_location:
            # Set the last found location as the current one (magenta).
            zones[current_location]['color'] = 'magenta'
            zones[current_location]['current'] = True
            print(f"Current Location: {current_location}")
        else:
            print("No current location found in Player.log.")

    except IOError as e:
        print(f"Error reading Player.log file {player_log_file}: {e}")

def add_locations_from_csv():
    """
    Reads known city/site locations from the 'cities.csv' file.
    Populates the global 'zones' dictionary with names and specific colors.
    CSV format: Location,Color,Name (e.g., 11.22.1.1.10,#554f97,Joppa)
    """
    filename = os.path.join(SAVE_DIR, LOCATIONS_CSV)

    if not os.path.exists(filename):
        print(f"Warning: Locations CSV file not found at {filename}")
        return

    # Regex to parse the CSV line: captures location, color, and name.
    csv_pattern = re.compile(r"^(\d{1,2}\.\d{1,2}\.\d\.\d\.\d{1,2}),(.+),(.+)$")

    try:
        with open(filename, 'r', encoding='utf-8') as f:
            for line in f:
                line = trim(line) # Trim leading/trailing whitespace from the line
                match = csv_pattern.match(line)
                if match:
                    zone_loc, color, name = match.groups()
                    zone_loc = trim(zone_loc)
                    color = trim(color)
                    name = trim(name)

                    zones[zone_loc] = zones.get(zone_loc, {}) # Ensure the inner dict exists
                    zones[zone_loc]['name'] = name
                    zones[zone_loc]['color'] = color
    except IOError as e:
        print(f"Error reading locations CSV file {filename}: {e}")

# --- HTML Generation Functions ---

def generate_html_header() -> str:
    """
    Generates the HTML <head> section with title, auto-refresh, and CSS styles.
    """
    return f"""<!DOCTYPE html>
<html>
<head>
    <title>CavesOfQud Parsang Map</title>
    <meta http-equiv="refresh" content="30">
    <style>
        table {{
            font-family: arial, sans-serif;
            border-collapse: collapse;
            width: 100%;
        }}
        th {{
            background-color: lightslategrey;
            border: 4px solid #d70513;
            position: sticky;
            top: 0;
            left: 0;
            text-align: center;
            padding: 8px;
        }}
        td {{
            border: 1px solid #dddddd;
            text-align: center;
            text-overflow: ellipsis;
            padding: 8px;
            white-space: nowrap; /* Prevent text wrapping in cells */
            overflow: hidden; /* Hide overflow */
        }}
        tr.border-bottom {{
            border-bottom: solid 4px #d70513;
        }}
        tr.border-top {{
            border-top: solid 4px #d70513;
        }}
        td.border-left {{
            border-left: solid 4px #d70513;
        }}
        td.border-right {{
            border-right: solid 4px #d70513;
        }}
        td.border-top {{
            border-top: solid 4px #d70513;
        }}
        td.border-bottom {{
            border-bottom: solid 4px #d70513;
        }}
        td.current-location {{
            border: solid 8px #f817d6;
        }}
    </style>
</head>
"""

def generate_html_table() -> str:
    """
    Generates the main HTML table content representing the map.
    """
    html_content = []
    html_content.append("<body>")
    html_content.append("\t<center><h1>Caves Of Qud Parsang Map</h1></center>")
    html_content.append(f"Generated: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    html_content.append("\t<table>")

    ZONE_DEPTH = 10
    PARSANG_X_MAX = 80
    PARSANG_Y_MAX = 25
    ZONE_DIM = 3 # 0, 1, 2 for x and y

    # Output map header row (Parsang X and Zone X labels)
    html_content.append("\t\t<tr>")
    html_content.append("\t\t\t<th>&nbsp;</th>") # Top-left empty cell
    for parsang_x in range(PARSANG_X_MAX):
        for zone_x in range(ZONE_DIM):
            html_content.append(f"<th>{parsang_x}.{zone_x}</th>")
    html_content.append("\t\t</tr>")

    # Output map data rows
    for parsang_y in range(PARSANG_Y_MAX):
        for zone_y in range(ZONE_DIM):
            tr_class = 'class="border-top"' if zone_y == 0 else ''
            html_content.append(f"\t\t<tr {tr_class}>")
            html_content.append(f"\t\t\t<th>{parsang_y}.{zone_y}</th>") # Row index (Parsang Y and Zone Y)

            for parsang_x in range(PARSANG_X_MAX):
                for zone_x in range(ZONE_DIM):
                    td_classes = []
                    if zone_x == 0:
                        td_classes.append("border-left")

                    zone_loc = f"{parsang_x}.{parsang_y}.{zone_x}.{zone_y}.{ZONE_DEPTH}"
                    zone_data = zones.get(zone_loc, {})

                    td_str = zone_data.get('name', '&nbsp;')
                    td_bgcolor = ''

                    if 'color' in zone_data:
                        if zone_data.get('current'):
                            td_classes.append("current-location")
                            # If it's the current location and has no name, show coordinates.
                            # The Perl script commented this out, so we'll mirror that.
                            # if td_str == '&nbsp;':
                            #     td_str = f"{parsang_x}.{zone_x} {parsang_y}.{zone_y}"
                        td_bgcolor = f"bgcolor=\"{zone_data['color']}\""

                    td_class_str = f"class=\"{' '.join(td_classes)}\"" if td_classes else ''
                    html_content.append(f"<td {td_class_str} {td_bgcolor}>{td_str}</td>")
            html_content.append("\t\t</tr>")

    html_content.append("\t</table>")
    html_content.append("</body>")
    return "\n".join(html_content)

def generate_html_footer() -> str:
    """
    Generates the HTML </html> closing tag.
    """
    return "</html>"

def generate_html_output():
    """
    Combines header, table, and footer to create the complete HTML map file.
    """
    try:
        with open(HTML_FILE, 'w', encoding='utf-8') as f:
            f.write(generate_html_header())
            f.write(generate_html_table())
            f.write(generate_html_footer())
        print(f"HTML map generated successfully to {HTML_FILE}")
    except IOError as e:
        print(f"Error writing HTML file {HTML_FILE}: {e}")

# --- Main Program Loop ---

def main_loop():
    """
    The main execution loop that monitors the Player.log file
    and regenerates the HTML map when changes are detected.
    """
    last_modified_time = 0
    # The Perl script had `read_cache_dir()` commented out, so we will too.
    # If you want to initially load zones from the cache, uncomment the line below.
    # read_zone_cache_dir()

    # The Perl script also had `add_locations()` commented out in the loop,
    # but it's common to load these at startup. Let's load them once here.
    add_locations_from_csv()

    while True:
        player_log_path = os.path.join(SAVE_DIR, "Player.log")

        if not os.path.exists(player_log_path):
            print(f"Waiting for Player.log to exist at {player_log_path}...")
            time.sleep(10)
            continue

        try:
            current_modified_time = os.path.getmtime(player_log_path)
            # Compare modification times. The Perl script used a difference of 15 seconds.
            # We'll use a simple inequality for more responsiveness.
            if current_modified_time != last_modified_time:
                print(f"Detected file change in Player.log... Regenerating map {current_modified_time}")

                read_player_log()
                # If you want to re-read cities.csv on every change (unlikely to be needed), uncomment below:
                # add_locations_from_csv()
                generate_html_output()

                last_modified_time = current_modified_time
            else:
                # print("No change detected in Player.log. Waiting...") # Optional: uncomment for more verbose output
                pass # No change, just sleep

        except OSError as e:
            print(f"Error accessing Player.log at {player_log_path}: {e}")

        # Limit disk thrashing and CPU usage.
        time.sleep(5) # Sleep for 10 seconds

# --- Entry Point ---
if __name__ == "__main__":
    main_loop()
